### PR TITLE
Split arch detection into separate cmake module

### PR DESCRIPTION
**Pull request by Dead2**
-  Add archdetect as separate cmake include file.
 - Simplify architecture checking by setting BASEARCH in addition to ARCH.
 - Clean up parts of CMakeLists.txt
 - Unify detection of ARM getauxval code availability.
    We don't want to compile arch-specific code when WITH_OPTIM is not set,
    and the current checks don't take that into account.
 -  Configure: Don't enable optimizations if without-optimizations has been set.
    Add ARM_GETAUXVAL define for ARM platforms if optimizations is enabled.

This also fixes a problem where gcc ppc64le builds sometimes attempt to build with -msse2 etc flags.

### CODE REVIEW

File: cmake/archdetect.c [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+// PowerPC
+#elif defined(__powerpc__) || defined(_ppc__) || defined(__PPC__)
+    #if defined(__64BIT__) || defined(__powerpc64__) || defined(__ppc64__)
+        #error archfound ppc64
+    #else
+        #error archfound ppc
+    #endif
+
+// --------------- Less common architectures alphabetically below ---------------
+#elif defined() || defined()
```
Comment (mtl1979) [inline cmake/archdetect.c]: This doesn't look right...

Comment (Dead2) [reply cmake/archdetect.c]: heh, no :)

Comment (Dead2) [reply cmake/archdetect.c]: Fixed

File: cmake/archdetect.cmake [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+# Make sure we have ARCH set
+if(NOT ARCH OR ARCH STREQUAL "unknown")
+    set(ARCH ${CMAKE_SYSTEM_PROCESSOR})
+    message(STATUS "Arch not recognized, falling back to cmake arch: '${ARCH}'")
+else()
+    message(STATUS "Arch detected: '${ARCH}'")
+endif()
+
+
+# Base arch detecion
```
Comment (nmoinvaz) [inline cmake/archdetect.cmake]: Spelling

File: cmake/archdetect.c [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+// PARISC
+#elif defined(__hppa__)
+    #error archfound parisc
+
+// RS-6000
+#elif defined(__THW_RS6000)
+    #error archfound rs6000
+
+// return 'unrecognized' if we do nok know what architecture this is
```
Comment (nmoinvaz) [inline cmake/archdetect.c]: Spelling

Comment (nmoinvaz) [inline cmake/archdetect.c]: Spelling

File: cmake/archdetect.cmake (hunk 2 above)
Comment (nmoinvaz) [inline cmake/archdetect.cmake]: Spelling

File: cmake/archdetect.cmake [hunk 4]
```diff
@@ -0,0 +1,77 @@
+# archdetect.c -- Detect compiler architecture and set ARCH and BASEARCH
```
Comment (nmoinvaz) [inline cmake/archdetect.cmake]: Should be archdetect.cmake not archdetect.c

File: cmake/archdetect.c (hunk 3 above)
Comment (Dead2) [reply cmake/archdetect.c]: Fixed

File: cmake/archdetect.cmake (hunk 2 above)
Comment (Dead2) [reply cmake/archdetect.cmake]: Fixed

File: cmake/archdetect.cmake (hunk 4 above)
Comment (Dead2) [reply cmake/archdetect.cmake]: Fixed

**Comment by nmoinvaz** [comment]
Interesting use of c to detect the architecture. 

- Double quotes such as those around `"${ARCH}"` are no longer necessary when doing `MATCHES`. Maybe we can clean that all up in another PR.
- Consider setting variables instead of setting BASEARCH to a string.

So instead of:
```
elseif("${ARCH}" MATCHES "alpha")
    set(BASEARCH "alpha")
```
You'd do this:
```
elseif("${ARCH}" MATCHES "alpha")
    set(BASEARCH_ALPHA_FOUND TRUE)
```

Then in the CMakeLists.txt you can simply do:

```
if(BASEARCH_ALPHA_FOUND)
```

**Comment by Dead2** [comment]
@nmoinvaz Those are good points :)
Not having to do regex checking of strings all the time should also lead to a slightly faster cmake.

**Comment by nmoinvaz** [comment]
Thanks. Yeah it is pretty standard nomenclature for cmake for scripts/includes/functions to have a `*_FOUND` variable that is set.

It might make some sense to put this into this PR (maybe somewhere at the top of the CMakeLists.txt):
```
message(STATUS "Using CMake version ${CMAKE_VERSION}")
```

**Comment by nmoinvaz** [comment]
Overall this PR is much needed. Nice work.

**Comment by Dead2** [comment]
I think I'll set both, I was thinking the BASEARCH variable could also be useful in conjunction with setting ARCHDIR and such.

Regarding the *_FOUND variable, just want to clarify, you want a ARCHDETECT_FOUND variable, right?

I think we can get the version printout in there :)

Thanks for the detailed feedback, that helps a lot.

**Comment by nmoinvaz** [comment]
Good point about being used in ARCHDIR. Well, I was thinking one _FOUND variable for each architecture. So there would be BASEARCH_X86_FOUND, etc.. There could also be an additional ARCHDETECT_FOUND to determine if any architecture was even detected.

**Comment by Dead2** [comment]
Oh, absolutely, one _FOUND for each arch, that I already added localyl.
I was wondering whether you wanted an additional one to detect that the script was found, I'll add one in any case.

Pushed the changes now.

File: CMakeLists.txt [hunk 5]
```diff
@@ -429,90 +388,92 @@ if(MSVC)
...
-    #else
-        crc = __builtin_ia32_crc32qi(crc, c);
-    #endif
-        (void)crc;
-        return 0;
-    }"
-    HAVE_SSE42CRC_INTRIN
-)
-set(CMAKE_REQUIRED_FLAGS)
+if("${BASEARCH}" MATCHES "x86")
```
Comment (nmoinvaz) [inline CMakeLists.txt]: Should be BASEARCH_X86_FOUND?

File: CMakeLists.txt [hunk 6]
```diff
@@ -4,6 +4,7 @@ if(CMAKE_VERSION VERSION_LESS 3.12)
 else()
   cmake_policy(VERSION 3.5.1...3.13.2)
 endif()
+message(STATUS "Using CMake version ${CMAKE_VERSION}")
```
Comment (nmoinvaz) [inline CMakeLists.txt:7]: "CMake" capitalization inconsistent with toolchain message() below..

File: CMakeLists.txt [hunk 7]
```diff
@@ -101,7 +88,7 @@ option(WITH_OPTIM "Build with optimisation" ON)
 option(WITH_NEW_STRATEGIES "Use new strategies" ON)
 option(WITH_NATIVE_INSTRUCTIONS
     "Instruct the compiler to use the full instruction set on this host (gcc/clang -march=native)" OFF)
-if("${ARCH}" MATCHES "arm" OR "${ARCH}" MATCHES "aarch64")
+if(BASEARCH_ARM_FOUND)
     option(WITH_ACLE "Build with ACLE CRC" ON)
     option(WITH_NEON "Build with NEON intrinsics" ON)
 elseif("${ARCH}" MATCHES "s390x")
```
Comment (nmoinvaz) [inline CMakeLists.txt]: Consider `BASEARCH_S390_FOUND AND "${ARCH}" MATCHES "s390x"`.. for clarity and a bit of speed.

File: CMakeLists.txt [hunk 8]
```diff
@@ -142,29 +129,22 @@ if(${CMAKE_C_COMPILER} MATCHES "icc" OR ${CMAKE_C_COMPILER} MATCHES "icpc" OR ${
...
     # /Oi ?
-    if(NOT ${ARCH} MATCHES "AMD64")
-        set(SSE2FLAG "/arch:SSE2")
-    endif()
-    if("${ARCH}" MATCHES "arm")
+    if(BASEARCH_X86_FOUND)
+        if(NOT ${ARCH} MATCHES "x86_64")
+            set(SSE2FLAG "/arch:SSE2")
+        endif()
+    elseif("${ARCH}" MATCHES "arm")
```
Comment (nmoinvaz) [inline CMakeLists.txt]: Intended to be BASEARCH_ARM_FOUND? I'm not sure on this one.

**Comment by nmoinvaz** [comment]
Things are looking good.

File: CMakeLists.txt (hunk 5 above)
Comment (Dead2) [reply CMakeLists.txt]: mismerge, thanks

File: CMakeLists.txt (hunk 6 above)
Comment (Dead2) [reply CMakeLists.txt:7]: What is correct though?

File: CMakeLists.txt (hunk 7 above)
Comment (Dead2) [reply CMakeLists.txt]: S360, but yeah, probably a good idea for clarity

File: CMakeLists.txt (hunk 8 above)
Comment (Dead2) [reply CMakeLists.txt]: Another mismerge, thanks for finding those :)

File: CMakeLists.txt (hunk 6 above)
Comment (nmoinvaz) [reply CMakeLists.txt:7]: Hmm. Probably "CMake" since that is what they show on their website. So probably the toolchain message() needs to be fixed.

Comment (Dead2) [reply CMakeLists.txt:7]: Agreed.

File: CMakeLists.txt (hunk 5 above)
Comment (nmoinvaz) [reply CMakeLists.txt]: This one still doesn't show yet.

Comment (Dead2) [reply CMakeLists.txt]: Doh, I tried adding it to the wrong commit, no wonder it was confused. It's in now.

**Pull request merged**